### ISSUE TITLE
TESTERS WANTED: Improving batching and thread utilisation at the same time

### ISSUE THREAD

**Issue by Jeffail** [Jeffail/benthos#291]
I've struggled in the past to find a batching solution that is able to:

1. Be completely flexible (count, byte size, period)
2. Be consistent across all components
3. Satisfy at-least-once delivery guarantees in all cases

The [`batch` processor](https://docs.benthos.dev/processors/#batch) was my first solution and satisfied 2 and 3 (with caveats). Although it offered all the parameters needed to satisfy 1 it would only resolve them during message receipt and isn't capable of flushing a backlog during shutdown. That makes it inappropriate for some common use cases.

The [`memory` buffer](https://docs.benthos.dev/buffers/#memory) was recently added in order to satisfy 1 and 2 entirely, which it does. However, the memory buffer (like any buffer) diminishes delivery guarantees and therefore cannot be considered a general solution. It also prevents users from having batch policies specific to certain inputs.

For the last year or so I've been experimenting with input level parameters for satisfying 2 and 3. The [`kafka_balanced` input](https://docs.benthos.dev/inputs/#kafka_balanced), for example, has the parameter `max_batch_count`. This mechanism has potential since being at the input level it allows us to preserve delivery guarantees without limitations on what kind of batching we implement (1 and 3). However, the challenge would be making this mechanism consistent across _all inputs_.

I think I have a solution that satisfies all three requirements and also comes with the benefit of improving the user experience of thread utilisation.

## Adding Async Methods to Inputs

The first stage of my plan is already in progress. I'm opening up the implementations of _all_ inputs so that they allow [asynchronous acknowledgement](https://github.com/Jeffail/benthos/blob/master/lib/input/reader/interface.go#L67) of messages in transit, with the potential (but not requirement) that subsequent messages can be processed before that ack is made.

This is important as up until now each input makes the assumption that when an acknowledgement is made for a message it applies to _all prior messages_. This new API will allow Benthos to process messages from a single input in parallel, utilising multiple processing threads and outputs. It also makes it very easy to create a general component that wraps these inputs with a buffer implementation that satisfies our above requirements, the resulting component configs will look something like this:

```yaml
type: kafka_cg
kafka_cg:
  addresses:
  - localhost:9092
  batching:
    byte_size: 0
    condition:
      type: static
      static: false
    count: 1
    period: ""
  consumer_group: benthos_consumer_group
  topics:
  - benthos_stream
```

Some inputs fit this new model easier than the previous one. The ones that don't are usually limited as they are ALO but consume a single continuous feed of data that cannot be acked individually (`kafka`, `kinesis`). However, there are individual solutions possible for these inputs.

### The Headache

The reason why this transition hasn't happened before is because of one single component, the `batch` processor. The way in which the `batch` processor operates relies entirely on the feed it receives being one continuous stream of data from a single source. This is why it has always been recommended that the `batch` processor is placed within the input section of a config.

For some asynchronous inputs it will be important that a `batch` processor is never present in a pipeline. Therefore, we need to be able to detect when this is the case in order to fallback to old behaviour and warn users.

## Plan

My plan is flexible but will work in the following stages:

1. ~Implement async methods for all inputs where applicable. Whist this is happening I'll also be putting work into improving and expanding the integrated test suite. For some inputs the asynchronous version will be a separate component entirely. Note that these methods being implemented does not mean they are being _used_ at this stage.~
2. ~Implement work around for `batch` processor used with async inputs.~
3. ~Gradually roll out `batching` fields for _all_ inputs. This is our new standard batching method. The documentation for each component will be updated to state the `batch` processor is NOT allowed with this input. In reality if the `batch` processor is used the input will fallback to the old implementation in order to preserve backwards compatibility. If stage 2 was not successful and it isn't possible to reliably detect `batch` processors at construction time then instead the new async methods will only be used then a user configures batching.~
4. ~Add batching to the `broker` input in order to support batching for combined async inputs. There will need to be documentation explaining that some inputs potentially will not work with it.~
5. ~Once all inputs support `batching` fields the documentation for the `batch` processor will be updated with a deprecation notice. There will also be a warning log when it is used. The [`batching`](https://docs.benthos.dev/batching/) and [`pipelines`](https://docs.benthos.dev/pipeline/) documentation will also be updated/simplified. All tutorials and guides using the `batch` processor will also be updated.~
6. Unless a serious flaw has been found with the new batching and input mechanisms the next major release of Benthos (v4) will remove the `batch` processor entirely and all code related to working around it will be removed.

Stage 1 is currently in progress and almost complete. If anyone has any thoughts please post them here, I'm very interested in feedback at this stage. Also, please let me know if you are testing any of the new components/batching behaviours.

**Comment by Jeffail** [comment]
Implementations are done for all inputs now. Any input component that doesn't have a `batching` field supports the `broker` `batching` mechanism. We therefore now have a substantial batching strategy for all inputs and all inputs where asynchronous processing is possible now has it supported by default.

**Comment by Jeffail** [comment]
These changes are implemented and scheduled to release in v3.3.0. There's a release candidate out so I'm looking for testers: https://github.com/Jeffail/benthos/releases/tag/v3.3.0-rc1
